src/io/kqueue.rs

```diff
@@ -0,0 +1,566 @@
+    }
+}
+
+bitflags::bitflags! {
+    /// The flags for a virtual node event.
+    pub struct VnodeEvents : u32 {
+        /// The file was deleted.
+        const DELETE = c::NOTE_DELETE;
+
+        /// The file was written to.
+        const WRITE = c::NOTE_WRITE;
+
+        /// The file was extended.
+        const EXTEND = c::NOTE_EXTEND;
+
+        /// The file had its attributes changed.
+        const ATTRIBUTES = c::NOTE_ATTRIB;
+
+        /// The file was renamed.
+        const RENAME = c::NOTE_RENAME;
+
+        /// Access to the file was revoked.
+        const REVOKE = c::NOTE_REVOKE;
+    }
+}
+
+#[cfg(feature = "process")]
+bitflags::bitflags! {
+    /// The flags for a process event.
+    pub struct ProcessEvents : u32 {
+        /// The process exited.
+        const EXIT = c::NOTE_EXIT;
+
+        /// The process forked itself.
+        const FORK = c::NOTE_FORK;
+
+        /// The process executed a new process.
+        const EXEC = c::NOTE_EXEC;
+
+        /// TODO
+        const TRACK = c::NOTE_TRACK;
+    }
+}
+
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+bitflags::bitflags! {
+    /// The flags for a user event.
+    pub struct UserFlags : u32 {
+        /// Ignore the user input flags.
+        const NOINPUT = c::NOTE_FFNOP;
+
+        /// Bitwise AND fflags.
+        const AND = c::NOTE_FFAND;
+
+        /// Bitwise OR fflags.
+        const OR = c::NOTE_FFOR;
+
+        /// Copy fflags.
+        const COPY = c::NOTE_FFCOPY;
+
+        /// Control mask for operations.
+        const CTRLMASK = c::NOTE_FFCTRLMASK;
+
+        /// User defined flags for masks.
+        const UDFMASK = c::NOTE_FFLAGSMASK;
+
+        /// Trigger the event.
+        const TRIGGER = c::NOTE_TRIGGER;
+    }
+}
+
+/// User-defined flags.
+///
+/// Only the lower 24 bits are used in this struct.
+#[repr(transparent)]
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct UserDefinedFlags(u32);
+
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+impl UserDefinedFlags {
+    /// Create a new `UserDefinedFlags` from a `u32`.
+    pub fn new(flags: u32) -> Self {
+        Self(flags & EVFILT_USER_FLAGS)
+    }
+
+    /// Get the underlying `u32`.
+    pub fn get(self) -> u32 {
+        self.0
+    }
+}
+
+/// A `kqueue`.
+pub struct Kqueue {
+    /// The underlying `kqueue` file descriptor.
+    fd: OwnedFd,
+}
+
+impl Kqueue {
+    /// Create a new `Kqueue`.
+    pub fn new() -> io::Result<Self> {
+        syscalls::kqueue().map(|fd| Self { fd })
+    }
+
+    /// Wait for events on this queue.
+    pub fn kevent(
+        &self,
+        changelist: &[InEvent<'_>],
```

Comment:
I'm not super familiar with the `kqueue` API, but as I understand it, the changelist registers file descriptors which are associated with the kqueue file descriptor, which may outlive the `&self` and ``` `_ ``` lifetimes here, so this lifetime description doesn't fully describe the possible file-descriptor lifetimes.

Reply:
I believe that, if a file descriptor is `close`d while it's registered in `kqueue`, it's also automatically deregistered from `kqueue`, so it shouldn't be a soundness issue, at the very least.

Reply:
The same thing could be said about directories, socket listeners, socket connections, udp sockets, signalfd, eventfd, and other kinds of file descriptors too though. Should rustix strive to have static types for every "type" of file descriptor?

Currently, rustix's approach is to assume that the bulk of the Rust code in the world won't be using rustix directly. Abstractions like std, cap-std, etc. provide these kinds of static type distinctions, in `File`, `ReadDir`, `TcpListener`, and so on. With that, rustix then focuses on just being a simple low-level API that isn't opinionated about which file descriptors can be used with which system calls. Maybe someone wants to `dup2` a file descriptor over a `kqueue` file descriptor. That's awkward to do if there's a `Kqueue` wrapper struct. And perhaps there are other things users want to do, that we can't anticipate. So rustix so far has erred on the side of letting users pass any file descriptor to any OS system call, because it isn't memory-unsafe, and in practice it hasn't been observed to be a problem in practice, and maybe there's a use case that rustix doesn't anticipate which users *do* want to do.

Reply:
(Oops, that comment should have been in reply to the other thread about the `Kqueue` type.)

Reply:
Thanks for the explanation! I've changed it so that there is just a freestanding `kqueue` and `kevent` function pair.